crates/languages/src/python/highlights.scm

```diff
@@ -1,18 +1,28 @@
+; Identifier naming conventions; these "soft conventions" should stay at the top of the file as they're often overriden
+
+; CamelCase for classes
+((identifier) @type.class
+  (#match? @type.class "^_*[A-Z][A-Za-z0-9_]*$"))
+
+; ALL_CAPS for constants:
+((identifier) @constant
+  (#match? @constant "^_*[A-Z][A-Z0-9_]*$"))
+
 (attribute attribute: (identifier) @property)
 (type (identifier) @type)
+; Any identifiers within a type are in practice going to be type identifiers; this nested scan
+; is currently the only way to support cases such as `type MyList[T] = T | list[T]`
+(type (_ (identifier) @type))
+(type (_ (_ (identifier) @type)))
```

Comment:
@osiewicz Currently this PR is still in draft, but already in advance I feel that this approach here is incredibly silly. I am not sure if it is possible to apply a rule for an arbitrarily nested case such as this?
```python
type AnyTasks[T: ExecutableTask | Task] = T | _TaskList[T] | list[T | _TaskList[T]] | None
```
I feel like the problem boils down to not getting adequate output from the parser so we're having to do more complicated stuff in the highlighter.

Reply:
Hmm. I'll revisit your PR tomorrow, but off the top of my head that's not going to be easy to pull off cleanly. Tree-sitter doesn't support matches of arbitrary depth. 
Besides, I think we should fall back on [semantic highlighting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) for cases like these. While in principle the only valid kind of identifier in that position is a type, users code might be invalid. Thus, if possible, I'd rather have us tackle semantic tokens instead of handling these cases at the tree-sitter level.

Don't get me wrong, I really appreciate your recent efforts on Python highlighting. It's just that this PR runs into a wall that should be brought down, not chipped away at :P

Reply:
Yeah I agree this is too big of a hack around the actual issue so I think I'll remove this part for now and maybe open an issue for it instead.

Reply:
With the other adjustments my python project is 99.8% correctly highlighted though so its a good experience compared to a month ago already :)

Reply:
Also regarding semantic tokens I believe basedpyright already has support for this https://docs.basedpyright.com/latest/benefits-over-pyright/pylance-features/.
This would make it [another big feature](https://github.com/zed-industries/zed/issues/7296#issuecomment-2495505867) in support of basedpyright at this stage.